Bounty Completion for SIGHASH_FORKID-based two-way replay protection

BTCGReplayUS opened the pull request:
This pull request implements SIGHASH_FORKID-based two-way replay protection for BTCG in a way similar to the way that BCH implements it.  However, it uses a FORKID of 79 (for the atomic number of GOLD) and it does not implement the un-needed runtime checks for the presence of a hardfork, because BTCG doesn't have running nodes to slowly upgrade the way BCH does.

It also fixes a number of warnings that were part of the equihash implementation.

leto commented:
This looks like really nicely documented code, and fixes lots of int/size_t bugs, :+1: to a FORKID of 79 😸

StarbuckBG commented:
Please contact me (Martin Kuvandzhiev) or h4x on Slack.

dexX7 commented on src/crypto/equihash.cpp:
```diff
@@ -225,12 +225,12 @@ FullStepRow<WIDTH>::FullStepRow(const unsigned char* hashIn, size_t hInLen,
     assert(len+lenIndices <= W);
     assert(len-trim+(2*lenIndices) <= WIDTH);
-    for (int i = trim; i < len; i++)
+    for (size_t i = static_cast<size_t>(trim); i < len; i++)
```
No need for the `static_cast<size_t>()` here, as `trim` is already passed as `size_t`.

dexX7 commented on src/crypto/equihash.cpp:
```diff
@@ -506,7 +506,7 @@ bool Equihash<N,K>::OptimisedSolve(const eh_HashState& base_state,
     const eh_index soln_size { 1 << K };
     std::vector<std::shared_ptr<eh_trunc>> partialSolns;
-    int invalidCount = 0;
+    size_t invalidCount = 0;
```
Why not use `unsigned int` or so here?

dexX7 commented on src/bitcoin-tx.cpp:
```diff
@@ -482,19 +482,30 @@ static void MutateTxDelOutput(CMutableTransaction& tx, const std::string& strOut
     const char *flagStr;
     int flags;
-} sighashOptions[N_SIGHASH_OPTS] = {
+} 
```
Nit: extra space

dexX7 commented on src/bitcoin-tx.cpp:
```diff
@@ -482,19 +482,30 @@ static void MutateTxDelOutput(CMutableTransaction& tx, const std::string& strOut
+    {"ALL|FORKID", SIGHASH_ALL | SIGHASH_FORKID},
+    {"NONE|FORKID", SIGHASH_NONE | SIGHASH_FORKID},
+    {"SINGLE|FORKID", SIGHASH_SINGLE | SIGHASH_FORKID},
+    {"ALL|FORKID|ANYONECANPAY",
```
Imho it's better readable, if it's on one line. Same applies to `core_write.cpp`.

idemin replied:
agree

dexX7 commented on src/test/script_tests.cpp:
```diff
@@ -1209,23 +1209,23 @@ BOOST_AUTO_TEST_CASE(script_combineSigs)
     scriptPubKey = GetScriptForMultisig(2, pubkeys);
     keystore.AddCScript(scriptPubKey);
-    SignSignature(keystore, txFrom, txTo, 0, SIGHASH_ALL);
+    SignSignature(keystore, txFrom, txTo, 0,SIGHASH_FORKID|SIGHASH_ALL);
```
Why remove those spaces? Looks ugly imho.

dexX7 commented on src/test/sighash_tests.cpp:
```diff
@@ -132,6 +132,7 @@ BOOST_AUTO_TEST_CASE(sighash_test)
     #endif
     for (int i=0; i<nRandomTests; i++) {
         int nHashType = InsecureRand32();
+ 	nHashType &= ~SIGHASH_FORKID;
```
Missing space.

dexX7 commented on src/test/txvalidationcache_tests.cpp:
```diff
@@ -184,10 +184,10 @@ BOOST_FIXTURE_TEST_CASE(checkinputs_test, TestChain100Setup)
         BOOST_CHECK(coinbaseKey.Sign(hash, vchSig));
         vchSig.push_back((unsigned char) 0); // padding byte makes this non-DER
-        vchSig.push_back((unsigned char)SIGHASH_ALL);
+        vchSig.push_back((unsigned char)(SIGHASH_ALL|SIGHASH_FORKID));
```
I'm wondering: given that it's always mandatory, why not update the value for the old sighash flags instead of combining both all the time?

BTCGReplayUS replied:
It's also just a maintainability nightmare to combine flags and have them mean different things on different implementations from what they are understood and documented to mean.

h4x3rotab commented:
This solution is roughly the same as another onging PR: https://github.com/BTCGPU/BTCGPU/pull/55

Just had a glimpse of the code. I think it's overall a great implementation. The major difference between BTG and BCH is that the first one won't need to care about pre-fork situation. So code can be simplified.

I'm going to review the code in detail. My first idea is that basic test cases are still needed. Though you've fixed the broken unit tests, it seems the full replay protection logic hasn't been covered. BCH has some good regression tests that can be used as a reference.

Keep going! And feel free to collaborate with us if you love. We are looking forward to the time when you complete the bounty.

h4x3rotab commented:
Failing unit test:
- test/multisig_tests.cpp
- test/transaction_tests.cpp

h4x3rotab commented on src/script/interpreter.cpp:
```diff
@@ -1173,8 +1194,20 @@ PrecomputedTransactionData::PrecomputedTransactionData(const CTransaction& txTo)
+    uint32_t nForkSigHashType=nHashType; // see item 10 in https://github.com/Bitcoin-UAHF/spec/blob/master/replay-protected-sighash.md
+    /* Two-way replay protection in BCH was implemented with a quadratic scaling fix replicating a lot of segwit's functionality.  They change the encoded versionbits inside the transaction hash (in addition to their
+       own 'segwit-lite' implementation.  Much/most of the replay-protected modified sighash in BCH is actually implementing that functionality.
+       However, because BTCG has segwit already, that fix is unnecessary.  The only thing necessary is to add the forkid to the nHashType in the hash to change the transaction hash.
```
However then most regtests will falling. Maybe we must set the fork height to 0, and do some random fixes.

jhoenicke replied:
It would be nice if forkid replay protected transactions always use SIGVERSION_WITNESS_V0 signatures even for non-segwit transactions, just like in BCH.  This makes it easier for hardware wallet especially if they already support BCH.

Is there a reason not to use SIGVERSION_WITNESS_V0 signatures for everything (and get rid of the code for old-style signatures)?

h4x3rotab commented on src/script/interpreter.cpp:
```diff
@@ -1258,7 +1291,10 @@ bool TransactionSignatureChecker::CheckSig(const std::vector<unsigned char>& vch
-    int nHashType = vchSig.back();
+
+    uint32_t nHashType = GetHashType(vchSig);
+    vchSig.pop_back();
```
Do you really want to pop it twice?

h4x3rotab commented:
After fix the double pop_back(), broken tests:

* test/script_tests.cpp
* test/sighash_tests.cpp

Failed regtests:
```
decodescript.py                | ✖ Failed  | 5 s
mempool_limit.py               | ✖ Failed  | 5 s
prioritise_transaction.py      | ✖ Failed  | 8 s
txn_clone.py                   | ✖ Failed  | 14 s
```

BTCGReplayUS commented:
Can't get access to your slack to discuss these changes, the slack invite from https://btcgpu.org is down.   I have made the changes and the new tests, but need to confirm something about the code live with @h4x3rotab.   Can he please post a UTC time when @h4x3rotab will be available?

h4x3rotab commented:
Please use this address: https://join.slack.com/t/bitcoin-gold/shared_invite/enQtMjYxMDU5NzQzNjUyLTJlODRhYjg2NGRmMzE4ZjBkZjlmOGM0ZDc0OGUyMDZmMTVlY2RjZDc2OWQ0NWY2NWI0ODcyZjYwZTVhMjQ5OTE

BTCGReplayUS commented:
I have made additional further commits.  Some of the cosmetic issues from equihash were fixed, but most importantly, I have resolved all the issues with the legacy sighash and script tests. 

All legacy tests now pass.

I do not have a testsuite implemented to test replay attacks, but I am working on that next, and it will only take a short time.

BTCGReplayUS commented:
Hello.  I have finished this and updated the pull request.  All legacy tests pass, all python regex tests pass, the code is commented, I fixed several small bugs, and there is an entire test suite inside src/test/script_tests.cpp to test the replay attack protection on all the permutations of successful transactions that can be implemented on BTC or on an other fork, including future forks that have both segwit and sighash_forkid enabled.

@h4x3rotab Thank you for your review and your pointing out of the bugs.

@StarbuckBG I detailed my preferred payment method in the open issues thread, however we can also discuss it here.  Thank you.

BTCGReplayUS commented:
I just enabled the blockheigh-dependant forkid check.

h4x3rotab commented:
Bounty was paid by BTC at Bitfinex rate 0.018BTC. You can confirm the payment here: https://blockchain.info/address/1Nphe82XxmoJSyWa5KgzsDEHV328oWWzxP

BTCGReplayUS commented:
I wanted to write here (and on the PR page) to confirm that with the code implemented and the bounty completed, the BTG team has followed through with payment of the bounty in full, on time, in the manner that I preferred.  This instills a significant amount of faith (to me) that the project is being run by people who have a deep sense of professionalism, and I am grateful to them for the opportunity.

Thank you very much.

jhoenicke commented on src/script/interpreter.cpp:
```diff
@@ -1173,8 +1195,18 @@ PrecomputedTransactionData::PrecomputedTransactionData(const CTransaction& txTo)
+       own 'segwit-lite' implementation.  Much/most of the replay-protected modified sighash in BCH is actually implementing that functionality.
+       However, because BTCG has segwit already, that fix is unnecessary.  The only thing necessary is to add the forkid to the nHashType in the hash to change the transaction hash.
+    */
+    if(nHashType & SIGHASH_FORKID)
```
It would be nice if forkid replay protected transactions always use SIGVERSION_WITNESS_V0 signatures even for non-segwit transactions, just like in BCH. This makes it easier for hardware wallet especially if they already support BCH.

Is there a reason not to use SIGVERSION_WITNESS_V0 signatures for everything (and get rid of the code for old-style signatures)?  Even Johnson Lau's forcenet suggested this for replay protection: https://github.com/jl2012/bitcoin/commit/7c2593946c4f3e210683110782d82f55473c682a#diff-be2905e2f5218ecdbe4e55637dac75f3L1188

h4x3rotab replied:
You are right. We are currently working on that but it's targeted for the next testnet launch (the testnet v1 will be reset to change the SIGHASH method)

saleemrashid replied:
@h4x3rotab To confirm: the mainnet will use `SIGVERSION_WITNESS_V0` for non-SegWit transactions?

h4x3rotab replied:
Yes. Track the progress here: https://github.com/BTCGPU/BTCGPU/pull/150

h4x3rotab commented:
Thank you, @BTCGReplayUS. Your implementation of replay protection is clear and effective. It helped us a lot. We've already merged our own version and will merge the full BIP143 style SIGHASH version very soon. So I'm going to close this PR.
